### PR TITLE
Queue application uplinks

### LINKED ISSUE #560: Queue ApplicationUp in NS while AS application link is unavailable
**Summary:** 

Replaces https://github.com/TheThingsNetwork/lorawan-stack/issues/439

NS should persistently queue ApplicationUp per application, while it is not available.

**Why do we need this?**

For the AS to recover from missed upstream events, including join-accepts with new session information, uplink messages as well as downlink queue invalidations with the new `FCntDown` to use in LoRaWAN 1.0.x.

**What is already there? What do you see now?**

NS drops upstream messages.
 
**What is missing? What do you want to see?**

A persistent queue per application UID. Once any AS instance links an application with the UID, NS flushes the queue to AS before it sends normal messages.

**How do you propose to implement this?**

Persistent queue in NS.

**Can you do this yourself and submit a Pull Request?**

Can review

### PR DESCRIPTION
**Pull request by rvolosatovs**
#### Summary

Closes #560 

#### Changes

- Queue application uplinks
- Add application uplink queue implementation using a persistent queue driven by Redis streams

#### Notes for Reviewers

Tried different approaches, but implementation with streams turned out to be the simplest and IMHO best.
Note that this implementation also allows for multiple NS instances to generate application uplinks concurrently, while only the instance holding the link will be sending them to the AS in proper order.

### CODE REVIEW

File: pkg/networkserver/grpc_asns.go [hunk 1]
```diff
@@ -49,51 +59,64 @@ func (ns *NetworkServer) LinkApplication(link ttnpb.AsNs_LinkApplicationServer)
...
+			default:
+				if err := link.Send(up); err != nil {
+					return err
+				}
+				_, err := link.Recv()
+				return err
+			}
+		}); err != nil {
+			logger.WithError(err).Debug("Close application link")
+			events.Publish(evtEndApplicationLink(ctx, ids, err))
```
[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/networkserver/grpc_asns.go]: It *can* be, but there's not reason to add such complexity here.
`LinkApplication` only returns when `Range` errors, hence that is also the only time when `EndApplicationLink` will be published and the payload is the error returned by `Range`

[the remark this answers is not in the document]
Comment (johanstokking) [reply pkg/networkserver/grpc_asns.go]: OK, I see now.

Why aren't we working with a channel between the registry and here? What is the upside of blocking on a read manually vs using a channel?

[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/networkserver/grpc_asns.go]: 1. Using channel here is inconsistent with other similar functionality and e.g. `sync.Map.Range` in stdlib
2. Channel requires an additional goroutine
3. Channel requires additional complexity(a wrapper struct is needed to hold context and uplink(s))
4. There's no reason to use a channel here, since we have to wait for `Range` to return completely anyway

What do you mean by "blocking manually"? `Range` blocks forever until `<-ctx.Done()` succeeds or error occurs

[the remark this answers is not in the document]
Comment (johanstokking) [reply pkg/networkserver/grpc_asns.go]: > 1. Using channel here is inconsistent with other similar functionality and e.g. `sync.Map.Range` in stdlib

That one, as well as range implementations of existing registries, are about existing items. This is a blocking consumer which is fundamentally different.

> 2. Channel requires an additional goroutine

Where?

Also, are you considering one goroutine per application link too expensive overhead?

> 3. Channel requires additional complexity - a wrapper struct is needed to hold context and uplink(s) and additional logic both here as well as in the interface implementations.

Yes, see also `ContextualApplicationUp` in Application Server.

> 4. There's no reason to use a channel here, since we have to wait for `Range` to return completely anyway

Why?

---

How about `Consume(..., callback)`, drop the `for { }`, and let the implementation continuously call the callback on new items?

Still, I think a channel (with contextual items), is cleaner.

[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/networkserver/grpc_asns.go]: > > 1. Using channel here is inconsistent with other similar functionality and e.g. `sync.Map.Range` in stdlib

> That one, as well as range implementations of existing registries, are about existing items. This is a blocking consumer which is fundamentally different.
Why so? The only difference from e.g. `sync.Map.Range`, here is that `sync.Map` is finite, whereas we're ranging over an infinite data structure.
This is modeled after `for x := range <channel>`.

> > 2. Channel requires an additional goroutine

> Where?

> Also, are you considering one goroutine per application link too expensive overhead?

There are different ways to "use a channel" here, if you provide me the method signature you are thinking about, I will provide you with the details.
Assuming the signature is `Range(<args>) (<-chan W, error)`, where `W` contains `*ttnpb.ApplicationUp`, `Context` and `error`, Redis implementation will have to `XReadGroup` in a goroutine in order to send downlinks on the channel.

a goroutine takes at least 2KB on Go 1.13
I think that if we can avoid using N*2KB extra memory space, where N is potentially quite large(N=application count in worst case), we should.
My biggest issue with having a goroutine is the added complexity of handling it. We would need additional `select` statements as well.

> > 4. There's no reason to use a channel here, since we have to wait for `Range` to return completely anyway

> Why?

https://github.com/TheThingsNetwork/lorawan-stack/pull/1524/files#diff-8f540de6ab7a805994f024884bbfc1e8R73

The simplicity of current solution is due to this assumption.



> How about Consume(..., callback), drop the for { }, and let the implementation continuously call the callback on new items?

> Still, I think a channel (with contextual items), is cleaner.

So current Redis implementation does *continuously call the callback on new items*, as well as the failed-to-process items. `for {}` block is not strictly necessary, it's there for 2 reasons:
1. Possible other implementations of `ApplicationUplinkQueue` may be represented as a finite data structure, in which case `Range` would return once the end is reached
2. Later we will add failure recovery and/or retrying of failed tasks, where we will have to call `Range` again.


Also, please note, that the actual callback need not be an anonymous function, and it may be configurable and have different implementations(e.g. `ns.applicationUplinks.Range(ctx, ids, ns.asUplinkHandler)`.
Also, the
```
			select {
			case <-ctx.Done():
				return ctx.Err()

			case <-ws.closeCh:
				cancel()
				<-ctx.Done()
				return ctx.Err()

			default:
```
block is, in fact, optional and is there just there to ensure failing quick and minimizing the chance of sending the uplink on the link after the context is done and/or link is closed - this is handled in goroutines, so there are no guarantees on the actual execution time, but it should be enough.
So, in fact the whole `Range` call can just look like this:
```
		if err := ns.applicationUplinks.Range(ctx, ids, func(ctx context.Context, up *ttnpb.ApplicationUp) error {
			if err := link.Send(up); err != nil {
				return err
			}
			_, err := link.Recv()
			return err
		}); err != nil {
			logger.WithError(err).Debug("Close application link")
			events.Publish(evtEndApplicationLink(ctx, ids, err))
			return err
		}
```


> Still, I think a channel (with contextual items), is cleaner.

Can you describe what you find "not clean" about current implementation? Channel-based implementation seems to be way more "dirty" to me in fact.

[the remark this answers is not in the document]
Comment (johanstokking) [reply pkg/networkserver/grpc_asns.go]: > Assuming the signature is `Range(<args>) (<-chan W, error)`, where `W` contains `*ttnpb.ApplicationUp`, `Context` and `error`, Redis implementation will have to `XReadGroup` in a goroutine in order to send downlinks on the channel.
> 
> a goroutine takes at least 2KB on Go 1.13
> I think that if we can avoid using N*2KB extra memory space, where N is potentially quite large(N=application count in worst case), we should.
> My biggest issue with having a goroutine is the added complexity of handling it. We would need additional `select` statements as well.

I care less about memory in this case and I don't see the complexity compared to what it looks like now.

> The simplicity of current solution is due to this assumption.

This is implementation-specific and not part of the interface.

> So current Redis implementation does _continuously call the callback on new items_, as well as the failed-to-process items

This is not defined in the interface either.

Also why is it called `Range()` and not `Consume()` already?

> 2\. Later we will add failure recovery and/or retrying of failed tasks, where we will have to call `Range` again.

OK

So basically `Range()` is the wrong term; `Consume()` or `Subscribe()` make more sense to me

[the remark this answers is not in the document]
Comment (rvolosatovs) [reply pkg/networkserver/grpc_asns.go]: Please see the updated version.
